.tools/validate.go

```diff
@@ -0,0 +1,196 @@
+		c, err := GitCommits(os.Getenv("TRAVIS_COMMIT_RANGE"))
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		for _, commit := range c {
+			fmt.Printf(" * %s %s ... ", commit.AbbreviatedCommit, commit.Subject)
+			if err := ValidateCommit(commit, DefaultRules); err != nil {
+				invalids = append(invalids, InvalidCommit{commit, err})
+				fmt.Printf("FAIL: %s\n", err)
+			} else {
+				fmt.Println("PASS")
+			}
+		}
+	}
+	if len(invalids) > 0 {
+		fmt.Printf("%d issues to fix\n", len(invalids))
+		os.Exit(1)
+	}
+}
+
+type ValidateRule func(CommitEntry) error
+
+func ValidateCommit(c CommitEntry, rules []ValidateRule) error {
+	for _, r := range rules {
+		if err := r(c); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type InvalidCommit struct {
+	CommitEntry CommitEntry
+	Error       error
+}
+
+type CommitEntry struct {
+	Commit               string
+	AbbreviatedCommit    string `json:"abbreviated_commit"`
+	Tree                 string
+	AbbreviatedTree      string `json:"abbreviated_tree"`
+	Parent               string
+	AbbreviatedParent    string `json:"abbreviated_parent"`
+	Refs                 string
+	Encoding             string
+	Subject              string
+	SanitizedSubjectLine string `json:"sanitized_subject_line"`
+	Body                 string
+	CommitNotes          string `json:"commit_notes"`
+	VerificationFlag     string `json:"verification_flag"`
+	ShortMsg             string
+	Signer               string
+	SignerKey            string `json:"signer_key"`
+	Author               Person `json:"author,omitempty"`
+	Commiter             Person `json:"commiter,omitempty"`
+}
+
+type Person struct {
+	Name  string `json:"name"`
+	Email string `json:"email"`
+	Date  string `json:"date"` // this could maybe be an actual time.Time
+}
+
+var (
+	prettyLogSubject = `--pretty=format:%s`
+	prettyLogBody    = `--pretty=format:%b`
+	prettyLogCommit  = `--pretty=format:%H`
+	prettyLogFormat  = `--pretty=format:{"commit": "%H", "abbreviated_commit": "%h", "tree": "%T", "abbreviated_tree": "%t", "parent": "%P", "abbreviated_parent": "%p", "refs": "%D", "encoding": "%e", "sanitized_subject_line": "%f", "commit_notes": "%N", "verification_flag": "%G?", "signer": "%GS", "signer_key": "%GK", "author": { "name": "%aN", "email": "%aE", "date": "%aD" }, "commiter": { "name": "%cN", "email": "%cE", "date": "%cD" }}`
+)
+
+func GitLogCommit(commit string) (*CommitEntry, error) {
+	buf := bytes.NewBuffer([]byte{})
+	cmd := exec.Command("git", "log", "-1", prettyLogFormat, commit)
+	cmd.Stdout = buf
+	cmd.Stderr = os.Stderr
+
+	if err := cmd.Run(); err != nil {
+		log.Println(strings.Join(cmd.Args, " "))
+		return nil, err
+	}
+	c := CommitEntry{}
+	output := buf.Bytes()
+	if err := json.Unmarshal(output, &c); err != nil {
+		fmt.Println(string(output))
+		return nil, err
+	}
+	output, err := exec.Command("git", "log", "-1", prettyLogSubject, commit).Output()
+	if err != nil {
+		return nil, err
+	}
+	c.Subject = strings.TrimSpace(string(output))
```

Comment:
On Wed, Sep 09, 2015 at 12:52:00PM -0700, Vincent Batts wrote:
> because subject and body can have symbols that --pretty does not
> escape. :-)

So can author names:

  $ git init
  $ git config user.name 'Mua"haha'
  $ touch README
  $ git add README
  $ git commit -m 'Testing crazy author name'
  $ git log --format='{"author": "%aN"}'
  {"author": "Mua"haha"}

which is why I'm suggesting “a loop over all user-supplied fields with
a single log call per-field/commit” [1] ;).

[1]: https://github.com/opencontainers/specs/pull/167/files#r39085317